appcelerator/titanium_mobile
# [TIMOB-26167] Android: Fixed TextField/TextArea focus/blur handling

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26172
- https://jira.appcelerator.org/browse/TIMOB-26167
- https://jira.appcelerator.org/browse/TIMOB-25819
- https://jira.appcelerator.org/browse/TIMOB-24138

**Summary:**
- [TIMOB-26167](https://jira.appcelerator.org/browse/TIMOB-26167) Fixed 7.3.0 regression (not released) where you can't edit TextField/TextArea within a ScrollView. (Caused by #9901 )
- [TIMOB-26172](https://jira.appcelerator.org/browse/TIMOB-26172) Fixed "focus" and "blur" events to not bubble up to parent views.
  * Now matches iOS' behavior and our documentation [here](https://docs.appcelerator.com/platform/latest/#!/guide/Event_Handling-section-src-29004896_EventHandling-Bubblingandnon-bubblingevents).
- Modified window to not set the focus to the first focusable child view when opened.
  * Resolves tickets [TIMOB-25819](https://jira.appcelerator.org/browse/TIMOB-25819) and [TIMOB-24138](https://jira.appcelerator.org/browse/TIMOB-24138).
  * Now matches iOS' behavior.

---
Testing TextField/TextArea/ScrollView handling:
1. Build and run the test code attached to [TIMOB-26167](https://jira.appcelerator.org/browse/TIMOB-26167) on Android.
2. On app startup, verify that the virtual keyboard does not pop-up. _(This verifies that a TextField/TextArea is no longer auto-focused on startup for [TIMOB-24138](https://jira.appcelerator.org/browse/TIMOB-24138).)_
3. Tap on a multiline TextArea.
4. Verify that you can edit the TextArea.
5. Scroll down to the very bottom.
6. Tap on a single-line TextField.
7. Verify that you can edit the TextField.

---
Testing "focus" and "blur" event bubbling:
1. Build and run the test code attached to [TIMOB-26172](https://jira.appcelerator.org/browse/TIMOB-26172) on Android.
2. Follow that ticket's test procedure, but this time verify the event's `bubbles` property is now `false`.
3. Verify that the window does not receive "focus" or "blur" events bubbled from the TextField. (Note that you will receive 1 "focus" event from the window on startup and this is normal.)

---
Testing `ScrollView` focus issue for [TIMOB-25819](https://jira.appcelerator.org/browse/TIMOB-25819):
1. Build and run the below code on Android.
2. Verify that the window is completely green. (If you see any red, then the test failed.)
3. Scroll down (by flinging up) and verify that you see a red view.

```javascript
var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
});
scrollView.add(Ti.UI.createView({
	backgroundColor: "green",
	width: Ti.UI.FILL,
	height: "100%",
}));
scrollView.add(Ti.UI.createListView({
	backgroundColor: "red",
	width: Ti.UI.FILL,
	height: "100%",
}));
window.add(scrollView);
window.open();
```
